kratos/spatial_containers/specialized_spatial_search.h

```diff
@@ -0,0 +1,807 @@
+    ///@{
+
+    Parameters mParameters; /// The configuration parameters
+
+    ///@}
+    ///@name Protected Operators
+    ///@{
+
+    ///@}
+    ///@name Protected Operations
+    ///@{
+
+    /**
+     * @brief This method returns the default parameters
+     * @return The default parameters
+     */
+    Parameters GetDefaultParameters() const;
+
+    ///@}
+    ///@name Protected  Access
+    ///@{
+
+    ///@}
+    ///@name Protected Inquiry
+    ///@{
+
+    ///@}
+    ///@name Protected LifeCycle
+    ///@{
+
+    ///@}
+private:
+    ///@name Static Member Variables
+    ///@{
+
+    ///@}
+    ///@name Member Variables
+    ///@{
+
+    ///@}
+    ///@name Private Operators
+    ///@{
+
+    ///@}
+    ///@name Private Operations
+    ///@{
+
+    /**
+     * @brief This method prepares the search
+     * @param rStructure The structure to be searched
+     * @param rInput The input to be searched
+     * @param rResults The results
+     * @param rResultsDistance The results distance
+     * @tparam TContainer The container type
+     * @tparam TResultType The result type
+     */
+    template<class TContainer, class TResultType>
+    std::vector<typename PointObject<typename TContainer::value_type>::Pointer> PrepareSearch(
+        const TContainer& rStructure,
+        const TContainer& rInput,
+        TResultType& rResults,
+        VectorDistanceType& rResultsDistance
+        )
+    {
+        // Some definitions
+        using ObjectType = typename TContainer::value_type;
+        using PointType = PointObject<ObjectType>;
+        using PointTypePointer = typename PointType::Pointer;
+        using PointVector = std::vector<PointTypePointer>;
+
+        // Creating the points
+        PointVector points;
+        const std::size_t structure_size = rStructure.size();
+        points.reserve(structure_size);
+        const auto it_begin = rStructure.begin();
+        for (std::size_t i = 0; i < structure_size; ++i) {
+            auto it = it_begin + i;
+            points.push_back(PointTypePointer(new PointType(*(it.base()))));
+        }
+
+        // Resizing the results
+        const std::size_t input_size = rInput.size();
+        if (rResults.size() != input_size) {
+            rResults.resize(input_size);
+        }
+        if (rResultsDistance.size() != input_size) {
+            rResultsDistance.resize(input_size);
+        }
+
+        return points;
+    }
+
+    /**
+     * @brief This method performs the search in parallel
+     * @param rInput The input container
+     * @param rRadius The radius array
+     * @param rSearch The spatial search
+     * @param rResults The results
+     * @param rResultsDistance The results distance
+     * @tparam TContainer The container type
+     * @tparam TSpatialContainer The spatial container type
+     * @tparam TResultType The result type
+     */
+    template<class TContainer, class TSpatialContainer, class TResultType>
+    void ParallelSearch(
```

Comment:
both `PrepareSearch` and `ParallelSearch` are `private` and only used in a single translation unit, so they'd be better off in the source file to reduce the weight of the header.

Reply:
They use a lot of templated arguments

Reply:
There are many template combinations, sorry

Reply:
Yes, but I need to do the explicit specializations and it is complicated with many combinations